### PR TITLE
SONARJS-701 Rule: Comma operator should not be used in switch-case expression

### PR DESCRIPTION
**Pull request by ivandalbosco**

### CODE REVIEW

File: javascript-checks/src/main/java/org/sonar/javascript/checks/NoCommaOperatorInSwitchCaseCheck.java [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+      addIssue(expression, getMessage());
+    }
+    
+    super.visitCaseClause(tree);
+  }
+
+  /**
+   * Gets the value of property "name" of annotation "@Rule" on this class.
+   * <p>
+   * TODO move this method to superclass
```
Comment (SonarTech) [inline javascript-checks/src/main/java/org/sonar/javascript/checks/NoCommaOperatorInSwitchCaseCheck.java]: INFO Complete the task associated to this TODO comment. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1135)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue
* INFO 1 info

Watch the comments in this conversation to review them.

File: javascript-checks/src/main/java/org/sonar/javascript/checks/NoCommaOperatorInSwitchCaseCheck.java [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+import org.sonar.squidbridge.annotations.SqaleConstantRemediation;
+
+@Rule(
+  key = "S3616",
+  name = "Do not use the comma operator within a switch case",
+  priority = Priority.MAJOR,
+  tags = {Tags.BUG})
+@ActivatedByDefault
+@SqaleConstantRemediation("5min")
+public class NoCommaOperatorInSwitchCaseCheck extends DoubleDispatchVisitorCheck {
```
Comment (vilchik-elena) [inline javascript-checks/src/main/java/org/sonar/javascript/checks/NoCommaOperatorInSwitchCaseCheck.java]: @ivandalbosco Please rename class to describe the bad pattern

File: javascript-checks/src/main/java/org/sonar/javascript/checks/NoCommaOperatorInSwitchCaseCheck.java [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+import org.sonar.plugins.javascript.api.tree.Tree.Kind;
+import org.sonar.plugins.javascript.api.tree.expression.ExpressionTree;
+import org.sonar.plugins.javascript.api.tree.statement.CaseClauseTree;
+import org.sonar.plugins.javascript.api.visitors.DoubleDispatchVisitorCheck;
+import org.sonar.squidbridge.annotations.ActivatedByDefault;
+import org.sonar.squidbridge.annotations.SqaleConstantRemediation;
+
+@Rule(
+  key = "S3616",
+  name = "Do not use the comma operator within a switch case",
```
Comment (vilchik-elena) [inline javascript-checks/src/main/java/org/sonar/javascript/checks/NoCommaOperatorInSwitchCaseCheck.java]: @ivandalbosco use title from RSPEC

File: javascript-checks/src/main/java/org/sonar/javascript/checks/NoCommaOperatorInSwitchCaseCheck.java [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+  priority = Priority.MAJOR,
```
Comment (vilchik-elena) [inline javascript-checks/src/main/java/org/sonar/javascript/checks/NoCommaOperatorInSwitchCaseCheck.java]: @ivandalbosco critical

File: javascript-checks/src/main/java/org/sonar/javascript/checks/NoCommaOperatorInSwitchCaseCheck.java [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+  tags = {Tags.BUG})
+@ActivatedByDefault
+@SqaleConstantRemediation("5min")
+public class NoCommaOperatorInSwitchCaseCheck extends DoubleDispatchVisitorCheck {
+  
+  @Override
+  public void visitCaseClause(CaseClauseTree tree) {
+    ExpressionTree expression = tree.expression();
+    if (expression.is(Kind.COMMA_OPERATOR)) {
+      addIssue(expression, getMessage());
```
Comment (vilchik-elena) [inline javascript-checks/src/main/java/org/sonar/javascript/checks/NoCommaOperatorInSwitchCaseCheck.java]: @ivandalbosco Use message from RSPEC (with number of operands and last operand)

File: javascript-frontend/src/main/java/org/sonar/plugins/javascript/api/visitors/DoubleDispatchVisitorCheck.java [hunk 6]
```diff
@@ -20,12 +20,13 @@
...
 
 import java.util.List;
+
 import org.sonar.javascript.visitors.Issues;
 import org.sonar.plugins.javascript.api.JavaScriptCheck;
 import org.sonar.plugins.javascript.api.tree.Tree;
 
 public abstract class DoubleDispatchVisitorCheck extends DoubleDispatchVisitor implements JavaScriptCheck {
-
+
```
Comment (vilchik-elena) [inline javascript-frontend/src/main/java/org/sonar/plugins/javascript/api/visitors/DoubleDispatchVisitorCheck.java]: @ivandalbosco Please avoid noop changes

File: javascript-checks/src/test/resources/checks/noCommaOperatorInSwitchCase.js [hunk 7]
```diff
@@ -0,0 +1,13 @@
+function foo() {
+  switch (a) {
+    case 1:         // OK
+      foo();
+    case 2,3:       // Noncompliant {{Do not use the comma operator within a switch case}}
```
Comment (vilchik-elena) [inline javascript-checks/src/test/resources/checks/noCommaOperatorInSwitchCase.js]: @ivandalbosco Please test issue location

File: javascript-checks/src/test/resources/checks/noCommaOperatorInSwitchCase.js [hunk 8]
```diff
@@ -0,0 +1,13 @@
+function foo() {
+  switch (a) {
+    case 1:         // OK
+      foo();
```
Comment (vilchik-elena) [inline javascript-checks/src/test/resources/checks/noCommaOperatorInSwitchCase.js]: @ivandalbosco please add `break`s

File: javascript-checks/src/main/java/org/sonar/javascript/checks/NoCommaOperatorInSwitchCaseCheck.java (hunk 2 above)
Comment (ivandalbosco) [inline javascript-checks/src/main/java/org/sonar/javascript/checks/NoCommaOperatorInSwitchCaseCheck.java]: Renamed to CommaOperatorInSwitchCaseCheck

File: javascript-checks/src/main/java/org/sonar/javascript/checks/NoCommaOperatorInSwitchCaseCheck.java (hunk 3 above)
Comment (ivandalbosco) [inline javascript-checks/src/main/java/org/sonar/javascript/checks/NoCommaOperatorInSwitchCaseCheck.java]: name changed to "Comma operators should not be used in switch cases"

File: javascript-checks/src/main/java/org/sonar/javascript/checks/NoCommaOperatorInSwitchCaseCheck.java (hunk 4 above)
Comment (ivandalbosco) [inline javascript-checks/src/main/java/org/sonar/javascript/checks/NoCommaOperatorInSwitchCaseCheck.java]: Done

File: javascript-checks/src/main/java/org/sonar/javascript/checks/NoCommaOperatorInSwitchCaseCheck.java (hunk 5 above)
Comment (ivandalbosco) [inline javascript-checks/src/main/java/org/sonar/javascript/checks/NoCommaOperatorInSwitchCaseCheck.java]: Done. Now using "Explicitly specify %d separate cases that fall through; currently this case clause only works for \"%s\"."

File: javascript-frontend/src/main/java/org/sonar/plugins/javascript/api/visitors/DoubleDispatchVisitorCheck.java (hunk 6 above)
Comment (ivandalbosco) [inline javascript-frontend/src/main/java/org/sonar/plugins/javascript/api/visitors/DoubleDispatchVisitorCheck.java]: Done

File: javascript-checks/src/test/resources/checks/noCommaOperatorInSwitchCase.js (hunk 8 above)
Comment (ivandalbosco) [inline javascript-checks/src/test/resources/checks/noCommaOperatorInSwitchCase.js]: Done

File: javascript-checks/src/main/java/org/sonar/javascript/checks/NoCommaOperatorInSwitchCaseCheck.java (hunk 1 above)
Comment (ivandalbosco) [inline javascript-checks/src/main/java/org/sonar/javascript/checks/NoCommaOperatorInSwitchCaseCheck.java]: No longer applicable (method removed)

File: javascript-checks/src/main/java/org/sonar/javascript/checks/NoCommaOperatorInSwitchCaseCheck.java (hunk 2 above)
Comment (ivandalbosco) [inline javascript-checks/src/main/java/org/sonar/javascript/checks/NoCommaOperatorInSwitchCaseCheck.java]: git ce5402a5

File: javascript-checks/src/main/java/org/sonar/javascript/checks/NoCommaOperatorInSwitchCaseCheck.java (hunk 3 above)
Comment (ivandalbosco) [inline javascript-checks/src/main/java/org/sonar/javascript/checks/NoCommaOperatorInSwitchCaseCheck.java]: git ce5402a5

File: javascript-checks/src/main/java/org/sonar/javascript/checks/NoCommaOperatorInSwitchCaseCheck.java (hunk 4 above)
Comment (ivandalbosco) [inline javascript-checks/src/main/java/org/sonar/javascript/checks/NoCommaOperatorInSwitchCaseCheck.java]: git ce5402a5

File: javascript-checks/src/main/java/org/sonar/javascript/checks/NoCommaOperatorInSwitchCaseCheck.java (hunk 5 above)
Comment (ivandalbosco) [inline javascript-checks/src/main/java/org/sonar/javascript/checks/NoCommaOperatorInSwitchCaseCheck.java]: git ce5402a5

File: javascript-frontend/src/main/java/org/sonar/plugins/javascript/api/visitors/DoubleDispatchVisitorCheck.java (hunk 6 above)
Comment (ivandalbosco) [inline javascript-frontend/src/main/java/org/sonar/plugins/javascript/api/visitors/DoubleDispatchVisitorCheck.java]: git ce5402a5

File: javascript-checks/src/test/resources/checks/noCommaOperatorInSwitchCase.js (hunk 7 above)
Comment (ivandalbosco) [inline javascript-checks/src/test/resources/checks/noCommaOperatorInSwitchCase.js]: Done
ce5402a5

File: javascript-checks/src/test/resources/checks/noCommaOperatorInSwitchCase.js (hunk 8 above)
Comment (ivandalbosco) [inline javascript-checks/src/test/resources/checks/noCommaOperatorInSwitchCase.js]: git ce5402a5

File: javascript-checks/src/main/java/org/sonar/javascript/checks/CommaOperatorInSwitchCaseCheck.java [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+  }
+  
+  /**
+   * Gets the number of "," characters in the specified expression.
+   * <p>
+   * Example 1: expression 10 (as in "case 10:") returns 0.
+   * <p>
+   * Example 2: expression 10,11,12,13 (as in "case 10,11,12,13:") returns 3.
+   */
+  private int getNumberOfCommas(ExpressionTree expression) {
```
Comment (SonarTech) [inline javascript-checks/src/main/java/org/sonar/javascript/checks/CommaOperatorInSwitchCaseCheck.java]: MAJOR Make "getNumberOfCommas" a "static" method. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS2325)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue
* MAJOR 1 major

Watch the comments in this conversation to review them.

File: javascript-checks/src/main/java/org/sonar/javascript/checks/CommaOperatorInSwitchCaseCheck.java [hunk 10]
```diff
@@ -30,30 +31,42 @@
...
+   * <p>
+   * Example 1: expression 10 (as in "case 10:") returns 0.
    * <p>
-   * TODO move this method to superclass
+   * Example 2: expression 10,11,12,13 (as in "case 10,11,12,13:") returns 3.
    */
-  private String getMessage() {
-    return getClass().getAnnotation(Rule.class).name();
+  private int getNumberOfCommas(ExpressionTree expression) {
+    int ret = 0;
```
Comment (vilchik-elena) [inline javascript-checks/src/main/java/org/sonar/javascript/checks/CommaOperatorInSwitchCaseCheck.java]: @ivandalbosco Please use more explicit name

File: javascript-checks/src/main/java/org/sonar/javascript/checks/CommaOperatorInSwitchCaseCheck.java [hunk 11]
```diff
@@ -30,30 +31,42 @@
...
+  
+  private static final String MESSAGE = "Explicitly specify %d separate cases that fall through; currently this case clause only works for \"%s\".";
   
   @Override
   public void visitCaseClause(CaseClauseTree tree) {
     ExpressionTree expression = tree.expression();
     if (expression.is(Kind.COMMA_OPERATOR)) {
-      addIssue(expression, getMessage());
+      int nbCommas = getNumberOfCommas(expression);
+      String lastCase = ((BinaryExpressionTree)expression).rightOperand().toString();
```
Comment (vilchik-elena) [inline javascript-checks/src/main/java/org/sonar/javascript/checks/CommaOperatorInSwitchCaseCheck.java]: @ivandalbosco please use `CheckUtils#asString` instread of `toString`

File: javascript-checks/src/main/java/org/sonar/javascript/checks/CommaOperatorInSwitchCaseCheck.java (hunk 10 above)
Comment (ivandalbosco) [inline javascript-checks/src/main/java/org/sonar/javascript/checks/CommaOperatorInSwitchCaseCheck.java]: Done in git 9400dc19

File: javascript-checks/src/main/java/org/sonar/javascript/checks/CommaOperatorInSwitchCaseCheck.java (hunk 11 above)
Comment (ivandalbosco) [inline javascript-checks/src/main/java/org/sonar/javascript/checks/CommaOperatorInSwitchCaseCheck.java]: Done in git 9400dc19

File: javascript-checks/src/main/java/org/sonar/javascript/checks/CommaOperatorInSwitchCaseCheck.java (hunk 10 above)
Comment (ivandalbosco) [inline javascript-checks/src/main/java/org/sonar/javascript/checks/CommaOperatorInSwitchCaseCheck.java]: Done in git dc3d9461

**Pull request merged**